### PR TITLE
Add I2C bus frequency readback property

### LINKED ISSUE #195: Add readback parameter for I2C baudrate
The `I2CMaster` class should store the configured baudrate in a `self.baudrate` variable so that sensors can check if the configured baudrate is compatible.

### PR DESCRIPTION
**Pull request by voldemort9999**
## Summary

While going through the codebase I noticed that `I2CMaster.configure()` sends the frequency to the hardware but doesn't keep track of what was set. This makes it impossible for sensor drivers to verify whether the bus speed is within their supported range — for example, the MLX90614 only supports up to 100 kHz SMBus, but there's currently no way to check this programmatically.

This PR adds a [frequency](cci:1://file:///c:/Users/BHAVESH/Documents/Code/Projects/gSoc/pslab-python/pslab/bus/i2c.py:456:4-459:30) property to [I2CMaster](cci:2://file:///c:/Users/BHAVESH/Documents/Code/Projects/gSoc/pslab-python/pslab/bus/i2c.py:439:0-492:20) that reflects the actual configured bus speed.

Closes #195

## Changes

- Added [_frequency](cci:1://file:///c:/Users/BHAVESH/Documents/Code/Projects/gSoc/pslab-python/pslab/bus/spi.py:72:4-78:44) attribute to `_I2CPrimitive.__init__` (initialized as `None`)
- [_configure()](cci:1://file:///c:/Users/BHAVESH/Documents/Code/Projects/gSoc/pslab-python/pslab/bus/i2c.py:66:4-90:31) now stores the actual hardware frequency after a successful configuration. The stored value is computed from the rounded [brgval](cci:1://file:///c:/Users/BHAVESH/Documents/Code/Projects/gSoc/pslab-python/pslab/sciencelab.py:305:4-307:56) register value (via [_get_i2c_frequency](cci:1://file:///c:/Users/BHAVESH/Documents/Code/Projects/gSoc/pslab-python/pslab/bus/i2c.py:98:4-100:66)), so it reflects what the hardware is actually running at rather than what was requested
- Added a read-only [frequency](cci:1://file:///c:/Users/BHAVESH/Documents/Code/Projects/gSoc/pslab-python/pslab/bus/i2c.py:456:4-459:30) property on [I2CMaster](cci:2://file:///c:/Users/BHAVESH/Documents/Code/Projects/gSoc/pslab-python/pslab/bus/i2c.py:439:0-492:20)

## Usage

    from pslab.bus.i2c import I2CMaster

    bus = I2CMaster()
    print(bus.frequency)      # default ~125 kHz

    bus.configure(4e5)
    print(bus.frequency)      # actual hardware frequency after rounding

## Tests

Added [tests/test_i2c_frequency.py](cci:7://file:///c:/Users/BHAVESH/Documents/Code/Projects/gSoc/pslab-python/tests/test_i2c_frequency.py:0:0-0:0) with 18 unit tests covering default frequency, reconfigurations, brgval rounding, read-only enforcement, boundary conditions, and error handling. All tests use `unittest.mock` and run without hardware.

    ============================= 18 passed in 0.05s =============================

## Summary by Sourcery

Add an I2C bus frequency readback on I2CMaster so code can query the actual configured bus speed.

New Features:
- Expose a read-only frequency property on I2CMaster that reports the current I2C bus speed in Hz.

Enhancements:
- Track the effective hardware I2C frequency inside the primitive I2C implementation after successful configuration so it can be reused by higher-level components.

Tests:
- Add comprehensive unit tests for I2C frequency tracking, including defaults, reconfiguration behavior, error cases, boundary conditions, and hardware interaction with a mocked device.

### CODE REVIEW

File: tests/test_i2c_frequency.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+        """Test with a high frequency whose brgval equals MIN_BRGVAL."""
+        # _get_i2c_frequency(MIN_BRGVAL) gives the theoretical max, but due
+        # to int() truncation in _get_i2c_brgval, that value may not
+        # round-trip. Instead, find a frequency that computes to MIN_BRGVAL.
+        max_brgval_freq = _I2CPrimitive._get_i2c_frequency(
+            _I2CPrimitive._MIN_BRGVAL + 1
+        )
+        master.configure(max_brgval_freq)
+        assert master.frequency is not None
+        assert master.frequency > 0
```
[the remark this answers is not in the document]
Comment (voldemort9999) [reply tests/test_i2c_frequency.py]: Fixed — added brgval round-trip verification to confirm the upper boundary is handled correctly.

File: tests/test_i2c_frequency.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+class TestFrequencyOnError:
+
+    def test_frequency_unchanged_on_invalid_frequency(self, master):
+        """If configure() raises ValueError, frequency should remain unchanged."""
+        master.configure(125e3)
+        original_freq = master.frequency
+
+        # Too low frequency should raise ValueError.
+        with pytest.raises(ValueError):
+            master.configure(1)  # Way too low
```
[the remark this answers is not in the document]
Comment (voldemort9999) [reply tests/test_i2c_frequency.py:142]: Good point. I've added test_frequency_unchanged_on_device_io_error which simulates an IOError during send_byte. The current _configure doesn't check get_ack() return value, so a NACK test isn't applicable here, but communication failures are now covered.

**Review by Saksham-Sirohi — COMMENTED** [review]
Copilot flagged the following:

- Note: `_frequency` stays `None` on `I2CSlave` instances that never call `configure()` — document that or set frequency when the slave triggers bus config.
